sudo/lib/system/term/ops.rs

```diff
@@ -0,0 +1,186 @@
+                    return Err(err);
+                }
+            }
+        }
+    }
+    // Restore the original action.
+    unsafe { sigaction(SIGTTOU, original_action.as_ptr(), std::ptr::null_mut()) };
+
+    Ok(())
+}
+
+/// Copy the settings of the `src` terminal to the `dst` terminal.
+pub fn term_copy<S: AsRawFd, D: AsRawFd>(src: &S, dst: &D) -> io::Result<()> {
+    let src = src.as_raw_fd();
+    let dst = dst.as_raw_fd();
+
+    let mut tt_src = MaybeUninit::<termios>::uninit();
+    let mut tt_dst = MaybeUninit::<termios>::uninit();
+    let mut wsize = MaybeUninit::<winsize>::uninit();
+
+    cerr(unsafe { tcgetattr(src, tt_src.as_mut_ptr()) })?;
+    cerr(unsafe { tcgetattr(dst, tt_dst.as_mut_ptr()) })?;
+
+    let tt_src = unsafe { tt_src.assume_init() };
+    let mut tt_dst = unsafe { tt_dst.assume_init() };
+
+    // Clear select input, output, control and local flags.
+    tt_dst.c_iflag &= !INPUT_FLAGS;
+    tt_dst.c_oflag &= !OUTPUT_FLAGS;
+    tt_dst.c_cflag &= !CONTROL_FLAGS;
+    tt_dst.c_lflag &= !LOCAL_FLAGS;
+
+    // Copy select input, output, control and local flags.
+    tt_dst.c_iflag |= tt_src.c_iflag & INPUT_FLAGS;
+    tt_dst.c_oflag |= tt_src.c_oflag & OUTPUT_FLAGS;
+    tt_dst.c_cflag |= tt_src.c_cflag & CONTROL_FLAGS;
+    tt_dst.c_lflag |= tt_src.c_lflag & LOCAL_FLAGS;
+
+    // Copy special chars from src verbatim.
+    tt_dst.c_cc.copy_from_slice(&tt_src.c_cc);
+
+    // Copy speed from src (zero output speed closes the connection).
+    {
+        let mut speed = unsafe { cfgetospeed(&tt_src) };
+        if speed == libc::B0 {
+            speed = libc::B38400;
+            unsafe { cfsetospeed(&mut tt_dst, speed) };
+            speed = unsafe { cfgetispeed(&tt_src) };
+            unsafe { cfsetispeed(&mut tt_dst, speed) };
+        }
+    }
+
+    tcsetattr_nobg(dst, TCSAFLUSH, &tt_dst)?;
+
+    cerr(unsafe { ioctl(src, TIOCGWINSZ, &mut wsize) })?;
+    cerr(unsafe { ioctl(dst, TIOCSWINSZ, &wsize) })?;
+
+    Ok(())
+}
+
+/// Set the `fd` terminal to raw mode. Enable terminal signals if `with_signals` is set to `true`.  
+pub fn term_raw<F: AsRawFd>(fd: &F, with_signals: bool) -> io::Result<()> {
+    let fd = fd.as_raw_fd();
+
+    if !unsafe { CHANGED } {
+        cerr(unsafe { tcgetattr(fd, OTERM.as_mut_ptr()) })?;
+    }
+    // Retrieve the original terminal.
+    let mut term = unsafe { OTERM.assume_init() };
+    // Set terminal to raw mode.
+    unsafe { cfmakeraw(&mut term) };
+    // Enable terminal signals.
+    if with_signals {
+        term.c_cflag |= ISIG;
+    }
+
+    tcsetattr_nobg(fd, TCSADRAIN, &mut term)?;
+    unsafe { CHANGED = true };
+
+    Ok(())
+}
+
+/// Restore the saved terminal settings if we are in the foreground process group.
+///
+/// This change is done after waiting for all the queued output to be written. To discard the
+/// queued input `flush` must be set to `true`.
+pub fn term_restore<F: AsRawFd>(fd: &F, flush: bool) -> io::Result<()> {
```

Comment:
this could potentially be misused if you call `term_raw(fd1)` and then `term_restore(fd2)` as there's no `OTERM` / `CHANGED` state per file descriptor. OTOH, that could also be totally fine because you can `dup` file descriptors and have different file descriptors (as in numbers) that refer to the same terminal / IO device ...

Reply:
Yeah... I'm tempted to refactor this into a `struct` that can be build from a single `OwnedFd` and move all those statics as fields of the `struct`.

That being said ogsudo only uses these over `/dev/tty` so it shouldn't be a problem.